### PR TITLE
refactor: stricter checks for spot price debugging in CL math

### PR DESCRIPTION
**Pull request by p0mvn**
## What is the purpose of the change

This PR refactors logic for retrieving spot price before and after CL quotes.

It makes it easier to reason about and debug.

Additionally, it adds some checks in the swap strategy to make sure that the next sqrt price is located in the right direction relative to the current given swap strategy. This helps catching errors early.

### ClickUp Task

[ClickUp Task URL](PASTE_CLICKUP_TASK_URL_HERE)

### CODE REVIEW

File: packages/math/src/pool/concentrated/one-for-zero.ts [hunk 1]
```diff
@@ -58,6 +58,8 @@ export class OneForZeroStrategy implements SwapStrategy {
     amountOutComputed: Dec;
     feeChargeTotal: Dec;
   } {
+    console.log("one for zero strat");
+    console.log("curSqrtPrice", curSqrtPrice.toString());
```
Comment (p0mvn) [inline packages/math/src/pool/concentrated/one-for-zero.ts:62]:
```suggestion
```

File: packages/pools/src/concentrated/pool.ts [hunk 2]
```diff
@@ -252,34 +251,38 @@ export class ConcentratedLiquidityPool implements BasePool, RoutablePool {
...
-      priceImpactTokenOut,
+      beforeSpotPriceInOverOut: beforeSpotPriceInOverOut.toDec(),
+      beforeSpotPriceOutOverIn: invertIfNonZero(
+        beforeSpotPriceInOverOut
+      ).toDec(),
+      afterSpotPriceInOverOut: afterSpotPriceInOverOut.toDec(),
+      afterSpotPriceOutOverIn: invertIfNonZero(afterSpotPriceInOverOut).toDec(),
+      effectivePriceInOverOut: effectivePriceInOverOut.toDec(),
+      effectivePriceOutOverIn: invertIfNonZero(effectivePriceInOverOut).toDec(),
+      priceImpactTokenOut: priceImpactTokenOut.toDec(),
```
Comment (p0mvn) [inline packages/pools/src/concentrated/pool.ts:285]: We will need to swtich to BigDec in v21 upgrade

File: packages/math/src/pool/concentrated/__tests__/quotes.spec.ts [hunk 3]
```diff
@@ -719,16 +719,18 @@ describe("calcOutGivenIn matches chain code", () => {
       ];
       const curSqrtPrice = new BigDec("70.710678118654752441");
       const swapFee = new Dec("0");
-      expect(
+      try {
```
Comment (jonator) [inline packages/math/src/pool/concentrated/__tests__/quotes.spec.ts:722]: FYI there's `expect().toThrow()` syntax for this

File: packages/math/src/pool/concentrated/one-for-zero.ts [hunk 4]
```diff
@@ -120,6 +120,15 @@ export class OneForZeroStrategy implements SwapStrategy {
       this.oneForZero.swapFee
     );
 
+    if (sqrtPriceNext.lt(curSqrtPrice)) {
+      throw new Error(
```
Comment (jonator) [inline packages/math/src/pool/concentrated/one-for-zero.ts:124]: I'd prefer string interpolation via ````

Comment (p0mvn) [reply packages/math/src/pool/concentrated/one-for-zero.ts:124]: OK

File: packages/math/src/pool/concentrated/__tests__/quotes.spec.ts (hunk 3 above)
Comment (p0mvn) [reply packages/math/src/pool/concentrated/__tests__/quotes.spec.ts]: Done

File: packages/pools/src/concentrated/pool.ts [hunk 5]
```diff
@@ -193,14 +193,15 @@ export class ConcentratedLiquidityPool implements BasePool, RoutablePool {
...
-      is0For1 ? tokenIn.denom : tokenOutDenom
-    );
+    const original_cur_sqrt_price = new BigDec(this.raw.current_sqrt_price);
+    const before1Over0SpotPrice = original_cur_sqrt_price.pow(new Int(2));
 
     /** Fetch ticks and calculate the out amount */
     let calcResult = undefined;
     do {
       const needMoreTicks = calcResult === "no-more-ticks";
+      // TODO: get current sqrt price from the return of this
```
Comment (p0mvn) [inline packages/pools/src/concentrated/pool.ts:203]: Next steps

**Review by jonator — APPROVED** [review]

**Pull request merged**